src/auto-resolve.ts

```diff
@@ -0,0 +1,136 @@
+      return (node.node.arguments[0] as any).value;
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Return array of dependencies
+ */
+function getDependencies(file: string, cwd: string) {
+  const code = fs.readFileSync(file, 'utf8');
+  if (!code) {
+    throw new Error(`Could not read ${file}`);
+  }
+  const ast = parse(code, {plugins: ['dynamicImport'], sourceType: 'module'});
+
+  const deps: string[] = [];
+
+  traverse(ast, {
+    enter(node) {
+      const moduleName =
+        importModuleName(node as NodePath<ImportDeclaration>) ||
+        dynamicImportModuleName(node as NodePath<CallExpression>);
+
+      if (!moduleName) {
+        return;
+      }
+
+      // if remote module
+      if (moduleName.startsWith('http://') || moduleName.startsWith('https://')) {
+        if (!deps.includes(moduleName)) {
+          deps.push(moduleName);
+        }
+        return;
+      }
+
+      try {
+        // if local module
+        const resolvedPath = path.resolve(path.dirname(file), moduleName);
+        if (exists(resolvedPath) && !deps.includes(resolvedPath)) {
+          deps.push(resolvedPath.replace(cwd, '')); // strip out full path above project root (cwd) when adding
+        }
+      } catch (e) {
+        // if npm module
+        if (!deps.includes(moduleName)) {
+          deps.push(moduleName);
+        }
+      }
+    },
+  });
+
+  return [...deps].sort((a, b) => a.localeCompare(b));
+}
+
+/**
+ * Check if filepath is valid
+ */
+
+function exists(filename: string) {
+  const stats = fs.statSync(filename);
+  return stats && stats.isFile();
+}
+
+/**
+ * Take an array of files, recursively scan imports, and return dependencies
+ */
+export default function autoResolve(files: string[], cwd: string, deps: DependencyTree = {}) {
+  files.forEach(file => {
+    const key = file.replace(cwd, ''); // don’t expose full filepath
+
+    // if file already parsed, skip
+    if (deps[key]) {
+      return;
+    }
+
+    // if new module, add to tree
+    const modules = getDependencies(file, cwd);
+    deps[key] = modules;
+
+    // if these modules can be resolved, try and resolve them
+    const localModules: string[] = [];
+    modules.forEach(moduleName => {
+      if (moduleName.startsWith('http://') || moduleName.startsWith('https://')) {
+        return; // skip remote modules
+      }
+
+      // if local, try and resolve
+      try {
+        const localModule = path.resolve(cwd, moduleName.replace(/^\//, ''));
+        if (exists(localModule) && !deps[localModule]) {
+          localModules.push(localModule); // only parse it if it’s a real file and it hasn’t been parsed
+        }
+      } catch (e) {
+        // don’t worry champ you tried your best (don’t do anything if file can’t be located)
+      }
+    });
+
+    // recursively scan these modules until there are none left
+    if (localModules.length) {
+      autoResolve(localModules, cwd, deps);
+    }
+  });
+
+  return deps;
+}
```

Comment:
Most of the stuff in this file isn’t too bad; it’s pretty reliable Babel parsing without any trickery. As the PR states, though, should Babel be used? Is there a way we can parallelize this to be faster?

Reply:
If you're not seeing a huge perf hit, then I think that this works to start. Since ESM is meant to be statically analyzable, a complex REGEX could theoretically do the job as well. I know I've tried this before with some success in another project: https://gist.github.com/manekinekko/7e58a17bc62a9be47172#gistcomment-2917250

Reply:
The common drawback I find to RegExs is comments. If I comment out `// import Preact from 'preact'`, that will be imported even when it shouldn’t. There are probably other edge cases, like only respecting top-level imports, but I’m not even sure this Babel plugin does that.

I feel safer using _some_ AST, mostly because I don’t trust my own RegExs. 😅

And yeah, from start to finish, I’m only seeing a few `ms` for the whole shebang. And this won’t get caught in an infinite parse loop or anything.

Reply:
+1, lets stick with this then 👍 I trust that Babel is pretty fast these days, especially for just parsing. There are also some other libraries only for AST parsing/walking that we could look at, but again I'd assume Babel's traverse library is pretty good :)